Fix for IndexView to allow AdminModel methods for fields #6076

Cheradenine opened the pull request:
Fixes issue #6076
For xlsx and csv export, list_export should behave like list_display. The bug was that the model_admin wasn't passed along to allow field lookups to be resolved by model_admin methods.

- Added method to do lookup on Book model to Author.date_of_birth
- Added tests to cover this case
- Verified that test fail
- Added fix in IndexView.get_heading
- Verified that tests now pass

jacobtoppm commented on wagtail/contrib/modeladmin/views.py:
```diff
@@ -283,7 +283,7 @@ def to_row_dict(self, item):
             if not value:
                 value = getattr(attr, 'empty_value_display', self.model_admin.get_empty_value_display(field))
-            row_dict[field] = value
+            row_dict[field] = force_str(value)
```
If you see https://github.com/wagtail/wagtail/blob/master/wagtail/admin/views/reports.py, `force_str` is currently chosen as the default in `get_preprocess_function` anyway, which allows the normal `SpreadsheetExportMixin` overrides (by value type and by field name) to be used, whereas adding it here would prevent those from taking place. I'd suggest removing it here.

Cheradenine commented:
Removed force_str - that makes sense. Stepped through the relevant code. Then I went down a rabbit hole with Excel and openpyxl where dates before 1900 are borked. They don't properly parse in Excel itself. So I added a comment, linked to the openpyxl issue. Goofy workaround to get tests to pass. Happy to discuss other options.

jacobtoppm commented:
> Removed force_str - that makes sense. Stepped through the relevant code. Then I went down a rabbit hole with Excel and openpyxl where dates before 1900 are borked. They don't properly parse in Excel itself. So I added a comment, linked to the openpyxl issue. Goofy workaround to get tests to pass. Happy to discuss other options.

Hmm, that was an interesting rabbit hole, thanks for catching that! The workaround functions for tests, but it seems that Excel won't properly display dates before 1900 anyway, so I think we need another option more generally eventually, and I'm not sure we want to be testing for this odd behaviour. For eventual solutions, one would be to replace

```
        (datetime.date, datetime.time): {FORMAT_XLSX: None},
```

in `SpreadsheetExportMixin.custom_value_preprocess` with a function that forces the date to a string if pre-1900. This isn't ideal, as if you had dates spanning the year 1900, some would be coerced to strings and some would remain in standard Excel date time format.

Another option would be to coerce all dates to strings. This is consistent, but means you don't have nice Excel dates automatically when exporting - especially relevant since most dates are likely to be post-1900 anyway!

A third would be to check all dates for pre-1900 values before outputting anything, and determining string vs Excel date format based on that.. This does result in an otherwise-unnecessary second loop through all data (as with things like modeladmin methods, we can't use a simple database query for the minimum), so isn't great for efficiency.

For now, until we make a further decision, I'd lean towards:
- Output in Excel datetime format as we do now (as the dates for builtin reports will all be post-1900, and for now we can encourage people to use `custom_field_preprocess` for custom reports)
- For tests: add an override to `custom_field_preprocess` on an `IndexView` to force the date to string, or to your date extracting method, for the purpose of testing the modeladmin methods for fields behaviour rather than this particular oddity
- We can then get this merged, and raise a separate issue for the 1900 date behaviour

Would you be happy to modify your tests slightly for now?

Cheradenine commented:
Moved the wonky date off by one code from the test, and as suggested we now use custom serialization for excel to iso format strings. This does seem cleaner - agree having that in the test was odd.
